Power Views specification

brew opened the pull request:
Please use this Pull Request for discussion related to the new CKAN 'PowerViews' extension. **DO NOT MERGE** until the spec has settled (likely when the feature is released).

brew committed:
Initial proposal for Power View extension

brew committed:
Update README with specs directory

brew committed:
Move powerviews spec into a directory

brew committed:
Related Work section

rufuspollock commented:
General comments (**in progress**):

To start with recommend re-reading: https://github.com/ckan/ckan/pull/1251#issuecomment-36668704

Core user stories (as i understand it):
- As a User I want to create visualizations (views) and save them so that I and others can see them in future
- As a User I want to create visualizations using data from several (resources) so that my visualization is richer
- (???) As a User I want create visualizations that incorporate multiple visual blocks
- As a User (or Org Owner) I want my visualization associated to me / my organization so that I get acknowledged, I can find it again, and permissions for editing etc are linked (so I can manage access and editing of visualizations from managing permissions for my organization)
- **As an CKAN "App" or Extension Developer I want to create new Dashboards / View Editors and allow users to customize and save the views/dashboards they create so that they are persisted**
  - I'm bolding this one because I think it is important. I anticipate this being by far the best way to allow people to quickly develop viz apps. Note we can even ignore "apps" that live out of the ckan domain - we can assume the app lives as an extension in CKAN but my point is that what it needs is access to an API (we do not need to build the viz wizard ourselves)
  - Doing this allows us to allow people to create view wizards in HTML+JS and just "plug them in"

## Implementation

The primary implication of the above is:

**Focus on building an API (forget the frontend) as the frontend (both creation and rendering) can be done by third parties (e.g. HDX frontend team).** This means implementation steps for CKAN extension in approx order are:
- working out authorization logic and ownership
- designing and implementing database tables
- logic layer (test driven)
- API (automatic from logic layer)

Once this is done I would focus on mocking out a pure HTML + JS viz creation (and rendering) wizard that persists over the API (rendering is included because you need preview normally anyway so you get rendering as part of creation). This can be entirely outside of the extension - it just needs to live on the domain and have access to relevant cookies or an API Key (so it can save over the API).

brew committed:
Changes to the scope and focus of PowerViews.

Focus is on an API for the creation of a PowerView object that provides
references to the one or more resources, and configured state.

brew committed:
Add API calls for adding and removing resources.

It's useful to be able to manage the resources associated with a
PowerView without having to update the whole PowerView dict.

brew committed:
More concise summary description

brew committed:
Swap name of powerview_resource_associate table

brew committed:
Add to db schema.

- private
- created_by
- owner_org
- created
- updated
